molgenis/molgenis-emx2
# feat(emx2) short time sorted auto-id using snowflake algorithm

Fixes https://github.com/molgenis/GCC/issues/183

**What are the main changes you did:**
- Implemented the Snowflake algorithm to create lexicographically(time) sortable and relatively 'short' ids.
- Changed it to fit our needs: short (base62 encoding) and with an instance identifier fur uniqueness across EMX instances.
- Added a database setting: "MOLGENIS_INSTANCE_ID", a random int (max 15 bits). Which is set at first start-up.
- Added a util function to retrieve the timestamp and instance id from the snowflake id.

Snowflake uses 64-bit by default, but this is arbitrary:
I chose the following configuration:
- 41 bits for the time component, resulting in 69 years of unique identifiers (2^41 ms)
- 8 bits for the sequence parts, this is the amount of ids you can generate per ms. So 8 bits = 256 ids per ms
- 15 bits left for the instance identifier: ~32K unique ids

**64** bits yields an id of max **11** character using the base62 characters (A-Za-z09).
Using less bits may result in a shorter id but not by much: you already need 7 for the time component.
If you're fine with creating 1 id per ms and you use 10 bits for the table id (1024 unique hashes) the maximum amount of characters is 9.

**how to test:**
- Look at the added unit test
- Create a table with auto-id and upload some records
- Add ${mg_autoid} to a computed value to add a post- or prefix.
- Or add records to the Order table in the pet store schema, and see the new autoid for the new records

**examples:**
Auto-id without prefix:
<img width="718" alt="image" src="https://github.com/user-attachments/assets/50a5b223-8fd5-42ac-813e-a8126e0bf117">

Computed value with prefix:
<img width="743" alt="image" src="https://github.com/user-attachments/assets/5b69994d-24f0-435d-ae5b-2201db80286f">





todo:
